modules/frontend/frontend.go

```diff
@@ -0,0 +1,227 @@
+package frontend
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"github.com/cortexproject/cortex/pkg/querier/frontend"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/weaveworks/common/user"
+	"net/http"
+)
+
+type FrontendConfig struct {
+	frontend.Config `yaml:",inline"`
+	ShardNum int    `yaml:"shard_num,omitempty"`
+}
+
+func (cfg *FrontendConfig) ApplyDefaults() {
+	cfg.Config.CompressResponses = false
+	cfg.Config.DownstreamURL = ""
+	cfg.Config.LogQueriesLongerThan = 0
+	cfg.Config.MaxOutstandingPerTenant = 100
+	cfg.ShardNum = 4
+}
+
+// NewTripperware returns a Tripperware configured with a middleware to split requests
+func NewTripperware(cfg FrontendConfig, logger log.Logger, registerer prometheus.Registerer) (frontend.Tripperware, error) {
+	level.Info(logger).Log("msg", "creating tripperware in query frontend to shard queries")
+	queriesPerTenant := promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Name:      "query_frontend_queries_total",
+		Help:      "Total queries sent per tenant.",
+	}, []string{"user"})
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		// get the http request, add some custom parameters to it, split it, and call downstream roundtripper
+		rt := NewRoundTripper(next, ShardingWare(cfg.ShardNum, logger, registerer))
+		return frontend.RoundTripFunc(func(r *http.Request) (*http.Response, error) {
+			level.Info(util.Logger).Log("msg", "request received by custom tripperware")
+			orgID := r.Header.Get(user.OrgIDHeaderName)
+			queriesPerTenant.WithLabelValues(orgID).Inc()
+
+			r = r.WithContext(user.InjectOrgID(r.Context(), orgID))
```

Comment:
shouldn't the org id already be in the context?  There is standard auth middleware that should do this for us and inject `single-tenant` if auth is disabled.